### ISSUE TITLE
4L 회고 / 23.07.20 / Day31  🦾✍💪

### ISSUE THREAD

**Issue by aaingyunii** [aaingyunii/Bootcamp_TIL#31]
## Liked
- `cifar10` 이미지 데이터셋을 이용해 직접 CNN 모델을 만드는 과정을 해볼 수 있어서 좋았습니다. 여러 가지 방법들을 사용하면서 어떻게 하면 더 개선될지에 대한 고민을 한 경험이 매우 좋았습니다. 

## Lacked
- TODO 학습에서 CNN 모델의 최고 성능을 구현하는 코드를 만들지 못해 아쉬웠습니다. 또한 아직도 모르는 내용들이 많다고 느껴져서 복습의 필요성을 다시 한번 느꼈습니다.

## Learned
- CNN 모델 네트워크 구현
Ex) 
```python
def get_mnist_cnn_model(lr=0.01):
    model = keras.Sequential()
    # Input Layer
    model.add(layers.InputLayer(input_shape=(28, 28, 1))) # 1 => Gray scale
    # Convolutional layer input data shape => 3rd array (height, width, channel)

    # Feature Extractor -> Convolution Block (Conv + MaxPool) 으로 정의
    # 영상용 Convolutional layer ==> .Conv2D() 사용
    model.add(layers.Conv2D(filters=16,    # filter(kernel) 의 개수. 1필터=1특징 찾음. 16필터=16특징
                           kernel_size=(3,3),
                           strides=(1,1), # default=1 / (좌우, 상하) 이동 크기, 같으면 정수로 설정. =1
                           padding='same', # padding='same' :입력, 출력의 size(h,w)가 같아지도록 padding 추가.
                                            # padding='valid' :(default) padding 추가하지 않는다.
                           activation='relu'
                           ))

    # .MaxPooling2D() :Feature map의 size를 줄인다. -> 지정한 크기의 영역에서 max값 하나를 추출.
    model.add(layers.MaxPooling2D(pool_size=(2, 2),
                                  strides=(2, 2), # default = None : pool_size와 동일한 크기
                                  padding="same" # "same" : 추출하려는 영역의 크기가 pool_size보다 작아도 max값 추출
                                            ## "valid" :(default)추출하려는 영역의 크기가 pool_size보다 작으면 추출X
                                  ))

    # filter 수를 늘리는 방향으로 정의.
    model.add(layers.Conv2D(filters=32, kernel_size=3, padding='same', activation='relu'))
    model.add(layers.MaxPooling2D(padding='same')) # pool_size 생략 => default :2 ==(2,2),
                                                    # strides == defalut pool_size랑 같다.

    model.add(layers.Conv2D(filters=64, kernel_size=3, padding='same', activation='relu'))
    model.add(layers.MaxPooling2D(padding='same'))
#-------------------Feature Extractor (Backbone network, Conv base)⬆


#-------------------추론기 ⬇----------------------------
    # Feature map 3차원 -> 1차원 변환, Dense layer 의 입력은 1차원만 가능하므로.
    model.add(layers.Flatten())

    model.add(layers.Dropout(0.3)) # 과적합, Overfitting 방지
    model.add(layers.Dense(units=256, activation='relu'))

    # 출력 Layer
    ## 다중분류 문제 : unit 수 - class의 개수(10), activation : softmax
    model.add(layers.Dropout(0.3))
    model.add(layers.Dense(units=10, activation='softmax', name='output_layer'))

    # 컴파일
    ## loss - categorical_crossentropy (y가 이미 one-hot encoding이 진행되었다면.)
    model.compile(optimizer=keras.optimizers.Adam(lr),
                  loss='categorical_crossentropy',
                  metrics=['accuracy'])

    return model
```

- 학습한 Weight (파라미터) 저장 및 불러오기
```
model.save_weights("저장경로")
model.load_weights('불러올경로')
```

- 전체 모델 저장하고 불러오기
```
model.save('저장할디렉토리')
tf.keras.models.load_model('저장된디렉토리')
```
- Callback을 사용한 모델 저장 및 Early Stopping
   - ModelCheckpoint
: 각 epoch 마다 학습한 모델과 weight(또는 weight만)를 저장한다. 지정한 평가지표(예:validation loss)가 가장 좋을 때 모델과 weight만 저장할 수 있다.
      - 주요 파라미터
: `save_weights_only=True` : True: 파라미터(weight)만 저장한다. False: 모델구조와 파라미터 모두 저장한다.
`save_best_only=True` : 학습중 성능이 개선될 때만 저장한다. (False:기본값 - 모든 에폭마다 저장한다.)
   - EarlyStopping
: Validation set에 대한 평가지표가 더 이상 개선되지 않을 때 학습을 자동으로 멈추는 callback
      - 주요 파라미터
: `monitor` : 모니터링할 평가지표 지정. (ex: accuracy)
patience: epoch 수 지정. validation 평가 지표가 개선이 안되더라도 지정한 epoch만큼 반복한다. 지정한 epoch만큼 반복 후에도 개선이 되지 않으면 중단한다.

## Longed for
- 복습!
